spatialmath/box.go

```diff
@@ -352,3 +355,95 @@ func separatingAxisTest(positionDelta, plane r3.Vector, halfSizeA, halfSizeB [3]
+// determines how many points to place per square mm. If the argument is set to 0. we
+// automatically substitute the value with defaultBoxPointDensity.
+func (b *box) ToPointCloud(resolution float64) []r3.Vector {
+	// check for user defined spacing
+	var iter float64
+	if resolution != 0. {
+		iter = resolution
+	} else {
+		iter = defaultBoxPointDensity
+	}
+
+	// TODO: the fillFaces method calls can be made concurrent if the ToPointCloud method is too slow
+	// the boolean values which are passed into the fillFaces method allow for common edges to be
+	// ignored. This removes duplicate points on the box's surface
+	var facePoints [][]r3.Vector
+	facePoints = append(facePoints, fillFaces(b.halfSize, iter, 0, true, false))
+	facePoints = append(facePoints, fillFaces(b.halfSize, iter, 1, true, true))
+	facePoints = append(facePoints, fillFaces(b.halfSize, iter, 2, false, false))
+
+	transformedVecs := transformPointsToPose(facePoints, b.Pose())
+	return transformedVecs
+}
+
+// fillFaces returns a list of vectors which lie on the surface of the box.
+func fillFaces(halfSize [3]float64, iter float64, fixedDimension int, orEquals1, orEquals2 bool) []r3.Vector {
+	var facePoints []r3.Vector
+	// create points on box faces with box centered at (0, 0, 0)
+	starts := [3]float64{0.0, 0.0, 0.0}
+	// depending on which faces we want to fill, we must keep either i xor j xor k constant
+	starts[fixedDimension] = halfSize[fixedDimension]
+	for i := starts[0]; lessThan(orEquals1, i, halfSize[0]); i += iter {
+		for j := starts[1]; lessThan(orEquals2, j, halfSize[1]); j += iter {
+			for k := starts[2]; k <= halfSize[2]; k += iter {
+				p1 := r3.Vector{i, j, k}
+				p2 := r3.Vector{i, j, -k}
+				p3 := r3.Vector{i, -j, k}
+				p4 := r3.Vector{i, -j, -k}
+				p5 := r3.Vector{-i, j, k}
+				p6 := r3.Vector{-i, j, -k}
+				p7 := r3.Vector{-i, -j, -k}
+				p8 := r3.Vector{-i, -j, k}
+
+				switch {
+				case i == 0.0 && j == 0.0:
+					facePoints = append(facePoints, p1, p2)
+				case j == 0.0 && k == 0.0:
+					facePoints = append(facePoints, p1, p5)
+				case i == 0.0 && k == 0.0:
+					facePoints = append(facePoints, p1, p7)
+				case i == 0.0:
+					facePoints = append(facePoints, p1, p2, p3, p4)
+				case j == 0.0:
+					facePoints = append(facePoints, p1, p2, p5, p6)
+				case k == 0.0:
+					facePoints = append(facePoints, p1, p3, p5, p8)
+				default:
+					facePoints = append(facePoints, p1, p2, p3, p4, p5, p6, p7, p8)
+				}
+			}
+		}
+	}
+	return facePoints
+}
+
+// lessThan checks if v1 <= v1 only if orEquals is true, otherwise we check if v1 < v2.
+func lessThan(orEquals bool, v1, v2 float64) bool {
+	switch {
+	case orEquals:
+		return v1 <= v2
+	default:
+		return v1 < v2
+	}
+}
+
+// transformPointsToPose gives vectors the proper orientation then translates them to the desired position.
+func transformPointsToPose(facePoints [][]r3.Vector, pose Pose) []r3.Vector {
+	var transformedVectors []r3.Vector
```

Comment:
Why a 2d array rather than 1d?

Reply:
can you please expand on this comment, I am unsure what it means..
I am guessing you are referring to why is `facePoints` of type `[][]r3.Vector`.
This is because `fillFaces` returns type `[]r3.Vector` which can only be appended to a list of type `[][]r3.Vector`.

Reply:
I see what you're saying.

This should take a 1d array, `[]r3.Vector`. I'll address how to deal with the append in a comment where it occurs.